### ISSUE TITLE
Click on tab bar to reload the same page?

### ISSUE THREAD

**Issue by saeropark** [react-community/react-navigation#1670]
Hi. 

I have 3 tabs: AM | PM | LUNCH 
 
the sequence for each tab is the same 
tab > loads page with list view > click on listview item > item details . 

My problem is, 
When a user is on AM tab and does not click 'back' to the listview page, the page is left hanging unless they go back. So assuming i did not do that and now i navigate on PM page and i want to click back on AM tab, i see the details and not the loaded page with listview. 

How do i go about such that when i am on other tabs, when i click back another tab i reload my listview instead of having to manually go back to reach the listview?

**Comment by augustoabreu** [comment]
Hi, isn't this behavior a little weird for users? I have a similar behavior but it only goes to "first page" when the user press a second time the tab item. Maybe could help you:

```
tabBarComponent:  ({jumpToIndex, ...props, navigation}) => (
    <TabBarBottom {...props} jumpToIndex={index => {
      const { state } = navigation,
            { routes } = state;
      if (state.index === index && routes[index].index !== 0) {
        navigation.dispatch(NavigationActions.reset({
    index,
    key: routes[index].key,
    actions: [
      NavigationActions.navigate({ routeName: routes[index].routeName })
    ]
  })
      } else {
        jumpToIndex(index);
      }
    }}/>
  )
```

**Comment by saeropark** [comment]
@augustoabreu  hi, `jumpToIndex(index)` meaning i have to set my screens as indexes under state? Currently my listview are nested inside a Stack.

**Comment by augustoabreu** [comment]
@saeropark 
The index in `jumpToIndex(index)` is the corresponding number of your tabs. Like this:

```
const tabs = TabNavigator({
  home: StackNavigator(), // index = 0
  users: StackNavigator(), // index = 1
  settings: StackNavigator(), // index = 2
})
```

The index in `state.index` is the current screen on `state.routes` array.

--

To  hide the header of main stack:

```
const stack = StackNavigator({
  login: StackNavigator,
  tabs: tabs
}, {
  headerMode: 'none'
})
```

**Comment by fabriziomoscon** [comment]
@augustoabreu I am using your code snippet but for some reason it doesn't work for tab1 and tab2. It works only for tab0. When I tap on Tab0 the stack is reset, but when I navigate inside tab2 stack and tap Tab2 I receive an error.

I am using:
`"react-navigation": "^1.0.0-beta.11"`
`"react": "16.0.0-alpha.12"`
`"react-native": "0.45.1"`
`"redux": "~3.6.0"`

cc @maxmcd

I have a StackNavigator containing a TabNavigator containing 3 StackNavigators.

```javascript
import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {
    Button,
    Platform,
} from 'react-native'
import {
    Header,
    StackNavigator,
    TabNavigator,
    TabBarTop,
    NavigationActions,
} from 'react-navigation'


const StackA = StackNavigator({
    ScreenA1: {
        screen: ScreenA1ViewContainer,
        navigationOptions: {
            header: null,
        },
    },
    ScreenA2: {screen: ScreenA2ViewContainer},
}, {
    navigationOptions: {
        tabBarLabel: 'TAB A',
    },
})

const StackB = StackNavigator({
    ScreenB1: {
        screen: ScreenB1ViewContainer,
        navigationOptions: {
            header: null,
        },
    },
    ScreenB2: {screen: ScreenB2ViewContainer},
}, {
    navigationOptions: {
        tabBarLabel: 'TAB B',
    },
})

// Root navigator is a StackNavigator
const StackC = StackNavigator({
    Account: {
        screen: ScreenC1ViewContainer,
        navigationOptions: {
            header: null,
        },
    },
    ScreenC2: {screen: ScreenC2ViewContainer},
}, {
    navigationOptions: {
        tabBarLabel: 'TAB C',
    },
})

// TabNavigator is nested inside StackNavigator
export const Tabs = TabNavigator({
    TabA: {screen: StackA},
    TabB: {screen: StackB},
    TabC: {screen: StackC},
}, {
    ...TabNavigator.Presets.AndroidTopTabs,
    tabBarPosition: 'top',
    animationEnabled: true,
    swipeEnabled: true,
    backBehavior: 'initialRoute',
    lazy: false,
    tabBarComponent:  ({jumpToIndex, ...props, navigation}) => (
        <TabBarTop
            {...props}
            jumpToIndex={tabIndex => {
                const { state } = navigation
                const { routes } = state
                if (routes[tabIndex].index !== 0) {
                    const stackRouteName = ['ScreenA1', 'ScreenB1', 'Account'][tabIndex]
                    navigation.dispatch(NavigationActions.reset({
                        index: tabIndex,
                        actions: [
                            NavigationActions.navigate({ routeName: stackRouteName })
                        ],
                    }))
                } else {
                    jumpToIndex(tabIndex)
                }
            }}
        />
    )
})

const OnboardingStack = StackNavigator({
    Onboarding: {screen: OnboardingViewContainer},
}, {
    headerMode: 'none',
})

const AppNavigator = StackNavigator({
    Home:       {screen: Tabs},
    Onboarding: {screen: OnboardingStack},
}, {
    headerMode: 'none',
})
```

I receive the error:
<img width="428" alt="screen shot 2017-06-16 at 11 03 05" src="https://user-images.githubusercontent.com/721890/27222702-3dc4220c-5285-11e7-884d-369fa02a7337.png">
<img width="845" alt="screen shot 2017-06-16 at 11 03 11" src="https://user-images.githubusercontent.com/721890/27222708-4318c4d8-5285-11e7-85c2-04dc3f1c21c0.png">


Any help would be greatly appreciated.

**Comment by shushuy** [comment]
@fabriziomoscon did you find a solution for the error?

**Comment by fabriziomoscon** [comment]
yes I solved.
I ended up using a fork:
`"react-navigation": "https://github.com/cribspot/react-navigation.git#187913749f61f27e109af841b5558235032652b7"`

check this out: https://github.com/react-community/react-navigation/pull/1335

**Comment by florianeckerstorfer** [comment]
I have the same issue as @fabriziomoscon. Is there any way to fix this problem or workaround that does not involve switching to a PR-branch of the project.

When someone can point me in the right direction, I can try to create a PR for this issue.

**Comment by beausmith** [comment]
@florianeckerstorfer The PR is already created. Check out [PR #1335](https://github.com/react-community/react-navigation/pull/1335)

**Comment by sanjayholla94** [comment]
@spencercarli I think this issue can be closed as tabBarOnPress is available in the latest release (1.0.0-beta.13)

**Comment by blaues0cke** [comment]
I came up with this solution to reset the active tab if its pressed another time while it is active:

```
tabBarComponent:  ({jumpToIndex, ...props, navigation}) => (
    <TabBarBottom
        {...props}
        jumpToIndex={tabIndex => {
            const currentIndex = navigation.state.index;

            if (tabIndex === currentIndex) {
                let type = null;

                switch (tabIndex) {
                    case 0: type = ActionTypes.RESET_TAB_QR_CODE;  break;
                    case 1: type = ActionTypes.RESET_TAB_HOME;     break;
                    case 2: type = ActionTypes.RESET_TAB_RECEIPTS; break;
                }

                navigation.dispatch({ type });
            }

            jumpToIndex(tabIndex);
        }}
    />
)
```

And in my tab navigation reducers I just do the following:

```
import * as   ActionTypes         from '../actions/types';
import      { NavigationActions } from 'react-navigation';
import        Registry            from '../lib/Registry';

const TabHomeStackNavigationReducer = (state, action) => {
    let copyiedAction = { ...action };

    if (action && state && action.type === ActionTypes.RESET_TAB_HOME) {
        copyiedAction.type    = NavigationActions.RESET;
        copyiedAction.index   = 0;
        copyiedAction.actions = [NavigationActions.navigate({ routeName: 'TabHome_Home' })];
    }

    const TabHomeStackNavigator = Registry.get('TabHomeStackNavigator');
    const newState              = TabHomeStackNavigator.router.getStateForAction(copyiedAction, state);

    return newState || state;
};

module.exports = TabHomeStackNavigationReducer;
```
